packages/publisher/s3/spec/PublisherS3.spec.ts

```diff
@@ -8,65 +8,81 @@ import {
   ForgeMakeResult,
   ResolvedForgeConfig,
 } from '@electron-forge/shared-types';
-import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
+import {
+  afterEach,
+  beforeEach,
+  describe,
+  expect,
+  it,
+  type Mock,
+  vi,
+} from 'vitest';
 
 import { PublisherS3, PublisherS3Config } from '../src/PublisherS3';
 
-// Mock AWS SDK modules
+// Get the actual fs module for test fixtures (before mocking)
+const actualFs = await vi.importActual<typeof import('node:fs')>('node:fs');
+
 vi.mock('@aws-sdk/client-s3');
 vi.mock('@aws-sdk/lib-storage');
 vi.mock('node:fs');
 
 describe('PublisherS3', () => {
   let publisher: PublisherS3;
-  let mockS3Client: vi.Mocked<S3Client>;
-  let mockUpload: vi.Mocked<Upload>;
+  let mockS3Client: S3Client;
+  let mockUploadOn: Mock;
+  let mockUploadDone: Mock;
   let tmpDir: string;
 
   beforeEach(async () => {
-    // Create temporary directory for test artifacts
     const tmp = os.tmpdir();
     const tmpdir = path.join(tmp, 'electron-forge-test-');
-    tmpDir = await fs.promises.mkdtemp(tmpdir);
+    tmpDir = await actualFs.promises.mkdtemp(tmpdir);
 
-    // Create test artifact files
-    await fs.promises.writeFile(
+    await actualFs.promises.writeFile(
       path.join(tmpDir, 'test-app-1.0.0.dmg'),
       'fake-dmg-content',
     );
-    await fs.promises.writeFile(
+    await actualFs.promises.writeFile(
       path.join(tmpDir, 'test-app-1.0.0.exe'),
       'fake-exe-content',
     );
-    await fs.promises.writeFile(
+    await actualFs.promises.writeFile(
       path.join(tmpDir, 'RELEASES'),
       'fake-releases-content',
     );
-    await fs.promises.writeFile(
+    await actualFs.promises.writeFile(
       path.join(tmpDir, 'RELEASES.json'),
       'fake-releases-json-content',
     );
 
-    // Mock S3Client constructor
     mockS3Client = {
       send: vi.fn(),
     } as any;
-    vi.mocked(S3Client).mockImplementation(() => mockS3Client);
 
-    // Mock Upload class
-    mockUpload = {
-      on: vi.fn().mockReturnThis(),
-      done: vi.fn().mockResolvedValue(undefined),
-    } as any;
-    vi.mocked(Upload).mockImplementation(() => mockUpload);
+    vi.mocked(S3Client).mockImplementation(function (this: S3Client) {
+      return mockS3Client;
+    } as unknown as typeof S3Client);
```

Comment:
nit: looks like the `as unknown` conversion is not necessary:


```suggestion
    });
```